### PR TITLE
thread/messages API improvements (was: add title to message thread json)

### PR DESCRIPTION
**Pull request by PetrDlouhy**
Some additions to the Message and Thread API.

### CODE REVIEW

**Comment by PetrDlouhy** [comment]
BTW @nikolai-b You haven't answered my question about the method similar to `as_json` for XML in #738. Aren't there some parameters exposed in XML format?

File: app/api/route/message_api.rb [hunk 1]
```diff
@@ -6,6 +6,8 @@ class MessageApi < Base
       optional :thread_id, type: Integer, desc: 'ID of thread'
       optional :order_by, type: Symbol, values:%i( created_at id  ), desc: 'Order of returned issues.'
       optional :order, type: Symbol, values:%i(asc desc), default: :asc
+      optional :id_lte, type: Integer, desc: 'Issue ID lower or equal than'
+      optional :id_gte, type: Integer, desc: 'Issue ID greater or equal than'
```
Comment (nikolai-b) [inline app/api/route/message_api.rb]: I'd prefer this done by date, have a `after` and `before` with a date then do this with `created_at`.  What do you think?  Have you already written code that expects `id_lte`?

**Comment by nikolai-b** [comment]
@PetrDlouhy very good point.  I hadn't realised that `default_format: :json` https://github.com/cyclestreets/cyclescape/blob/staging/app/api/api.rb#L7 allowed other formats.  Are you using XML?  My reaction is that we just lock it down and only expose JSON if that works for you.
If you need XML then I was wrong to suggest `as_json` and we need to use [Grape Entity](https://github.com/ruby-grape/grape-entity) which, although it makes sense, I'd prefer to avoid due to the overhead.

**Comment by nikolai-b** [comment]
@PetrDlouhy thanks, that is good news: 3a7e62ec7f67ddbac73c12c641080d0d1e09f27e
@mvl22 as this is a breaking change do you know if anyone is using the API in XML (or txt!)?  I can't see anything obvious in the logs but our API does respect the `Accept` header which wouldn't appear in the logs.

**Comment by mvl22** [comment]
I suggest that the whole API should be JSON only, unless an XML implementation is essentially trivial or automatic.

I don't think XML is being used, and XML popularity has plummeted while JSON has massively increased.

**Comment by mvl22** [comment]
Any output should always do the filtering before the output specification. In general I'd like a really clear list of what exactly is exposed via the API. Anything relating to individual user data must be tightly locked down, especially with GDPR changes coming.

**Comment by nikolai-b** [comment]
XML is trivial (the gem we are using does it automatically) but I'd prefer not to support it as we are sanitizing the responses.  Without them sanitized I can't see any private fields, just confusing ones (it simply returns the equivalent of the what is in the database e.g. https://github.com/cyclestreets/cyclescape/blob/staging/db/schema.rb#L350-L364
 
https://www.cyclescape.org/api is a clear list of what is exposed, in terms of issues and messages then only public ones are displayed.  I'm less sure about returning 403 forbidden as this is leaking information as what exists and what doesn't, I'd prefer 404 for any private resource or non-existent resource and this is what we currently do.

File: app/api/route/message_api.rb (hunk 1 above)
Comment (PetrDlouhy) [reply app/api/route/message_api.rb]: No, I can easily change it to `after_id` and `before_id`.

**Comment by PetrDlouhy** [comment]
@nikolai-b I have rewritten this to the `stard_date` and `end_date` parameters.

**Comment by PetrDlouhy** [comment]
I have added the `public_token` attributes to the message and thread JSONs, so I am able to create reply e-mail address.

**Comment by PetrDlouhy** [comment]
@nikolai-b Now I realised, that I need something a bit different, than is `start_date` and `end_date`. I need to keep something as bookmark to the last issue and get newer threads in the next run of my script. The difference is, that it needs to be in `DateTime` type and it needs to be only `less than`.

I added `datetime_less_than` parameter to illustrate that, but it can be renamed to something more coherent with rest of the application.

**Comment by mvl22** [comment]
> datetime_less_than

How about `datetime_until`

File: app/api/route/thread_api.rb [hunk 2]
```diff
@@ -7,6 +7,9 @@ class ThreadApi < Base
       optional :issue_id, type: Integer, desc: 'ID of issue'
       optional :order_by, type: Symbol, values:%i( created_at id  ), desc: 'Order of returned issues.'
       optional :order, type: Symbol, values:%i(asc desc), default: :asc
+      optional :end_date, type: Date, desc: 'No threads after the end date are returned'
+      optional :start_date, type: Date, desc: 'No threads before the start date are returned'
+      optional :datetime_less_than, type: DateTime, desc: 'No threads before the start date and time are returned'
```
Comment (nikolai-b) [inline app/api/route/thread_api.rb]: Can we just have 
```ruby
optional :before, type: [Date, DateTime], desc: : 'Only threads before this date are returned'
optional :after, type: [Date, DateTime], desc: : 'Only threads after or equal to this date are returned'
```

File: app/api/route/thread_api.rb [hunk 3]
```diff
@@ -19,6 +22,9 @@ def post_or_get_thread
         end
         scope = scope.order(params[:order_by] => params[:order]) if params[:order_by]
         scope = scope.where(issue_id: params[:issue_id]) if params[:issue_id]
+        scope = scope.before_date(params[:end_date]) if params[:end_date]
+        scope = scope.after_date(params[:start_date]) if params[:start_date]
+        scope = scope.after_date_not_equal(params[:datetime_less_than]) if params[:datetime_less_than]
```
Comment (nikolai-b) [inline app/api/route/thread_api.rb]: Same here.

File: app/models/message.rb [hunk 4]
```diff
@@ -73,6 +73,16 @@ class Message < ActiveRecord::Base
...
 
+  class << self
+    def after_date(date)
+      where('messages.created_at >= ?', date)
+    end
+
+    def before_date(date)
+      where('messages.created_at <= ?', date)
+    end
+  end
```
Comment (nikolai-b) [inline app/models/message.rb]: These can be `scope`

```ruby
scope :after_date, ->(date) { where(arel_table[created_at].gteq(date) }
scope :before_date, ->(date) { where(arel_table[created_at].lt(date) }
```

File: app/models/message_thread.rb [hunk 5]
```diff
@@ -104,6 +104,18 @@ class MessageThread < ActiveRecord::Base
...
+      where('message_threads.created_at >= ?', date)
+    end
+
+    def after_date_not_equal(date)
+      where('message_threads.created_at > ?', date)
+    end
+
+    def before_date(date)
+      where('message_threads.created_at <= ?', date)
+    end
```
Comment (nikolai-b) [inline app/models/message_thread.rb]: Same here scopes can be used here.

**Comment by nikolai-b** [comment]
@PetrDlouhy if your happy with these changes I can do them after merging or I can leave it to you.

**Comment by PetrDlouhy** [comment]
@nikolai-b I will make the changes myself and I will test, if it behaves like I need. So please wait with the merging.

File: app/api/route/thread_api.rb (hunk 2 above)
Comment (PetrDlouhy) [reply app/api/route/thread_api.rb]: I think, that the `[Date, DateTime]` doesn't work. I have got
```
{
  "error": "before is invalid"
}
```
for both date and datetime format on input.

Comment (PetrDlouhy) [reply app/api/route/thread_api.rb]: Aha, it is `types` not `type` in that case.

**Comment by nikolai-b** [comment]
@PetrDlouhy great thanks!

File: app/api/route/thread_api.rb (hunk 2 above)
Comment (nikolai-b) [reply app/api/route/thread_api.rb]: :disappointed: sorry about that, I should have checked myself first.

**Comment by PetrDlouhy** [comment]
@nikolai-b I have reworked this. I have chosen the `gt` operator for `after_date` instead of `qteq`, because I think it better fits the filter name and description.

There is problem with filtering by `DateTime`. The returned `created_at` parameter is rounded at 3 decimal positions. It means, that filtering by the returned values is quite undeterministic. If I get `created_at` from the newest thread and give it as `after_date` parameter, than the last issue can be in the list or can be missing. That makes it unusable in my use-case. Therefore I have added `after_id` parameter for threads.

**Comment by PetrDlouhy** [comment]
@nikolai-b I have realised other problem - other inconsistency with issue API. In issue API, the `created_at` is returned in milliseconds without time zone information. It is not possible to filter directly by values of that parameter.

**Comment by nikolai-b** [comment]
@PetrDlouhy ah, I see.
I think with issue it was a mistake to use `.to_i` so I've done 75251d49e604504da54e3b657e7dacb9bcb67c88 and with the resolution stuck at 3 decimal places I'd never noticed, I've bumped it with 1306a5dc987cbe47f1237ae386285678771933f3.  Does that help?
This MR is fine by me so I'm happy to merge it if your finished.
Thanks!

File: app/api/route/message_api.rb [hunk 6]
```diff
@@ -6,6 +6,8 @@ class MessageApi < Base
       optional :thread_id, type: Integer, desc: 'ID of thread'
       optional :order_by, type: Symbol, values:%i( created_at id  ), desc: 'Order of returned issues.'
       optional :order, type: Symbol, values:%i(asc desc), default: :asc
+      optional :after, types: [DateTime, Date], desc: 'Only messages after the end date or time are returned'
+      optional :before, types: [DateTime, Date], desc: 'Only messages before the start date or time are returned'
```
Comment (nikolai-b) [inline app/api/route/message_api.rb]: Sorry for suggesting `:after` and `:before` 
I hadn't noticed we were using `:end_date` and `:start_date` already so I think we should stick with that.
https://github.com/cyclestreets/cyclescape/blob/master/app/api/route/issue_api.rb#L12-L13

**Comment by PetrDlouhy** [comment]
@nikolai-b OK, nice. I have reworked this to copy issues API scheme. The params are `start_date`/`end_date` now and the inequality is `lteq` and `gteq`.

**Pull request merged**